File: 2025/01 January/13 Container With Most Water/main.py
# Link: https://www.geeksforgeeks.org/problems/container-with-most-water0535/1


class Solution:
    def maxWater(self, arr):
        n = len(arr)
        i, j = 0, n - 1
        maxi = 0
        while i < j:
            maxi = max(maxi, min(arr[i], arr[j]) * (j - i))

            if arr[i] <= arr[j]:
                i += 1
            else:
                j -= 1
        return maxi


if __name__ == "__main__":
    object = Solution()
    arr = [1, 5, 4, 3]
    print(object.maxWater(arr))

    arr = [3, 1, 2, 4, 5]
    print(object.maxWater(arr))

    arr = [2, 1, 8, 6, 4, 6, 5, 5]
    print(object.maxWater(arr))

File: 2025/01 January/13 Container With Most Water/main.cpp
// Link: https://www.geeksforgeeks.org/problems/container-with-most-water0535/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int maxWater(vector<int> &arr)
    {
        int n = arr.size();
        int i = 0, j = n - 1;
        int maxi = 0;
        while (i < j)
        {
            maxi = max(maxi, min(arr[i], arr[j]) * (j - i));

            if (arr[i] <= arr[j])
                i++;
            else
                j--;
        }
        return maxi;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {1, 5, 4, 3};
    cout << object.maxWater(arr) << endl;

    arr = {3, 1, 2, 4, 5};
    cout << object.maxWater(arr) << endl;

    arr = {2, 1, 8, 6, 4, 6, 5, 5};
    cout << object.maxWater(arr) << endl;
    return 0;
}